README.md
# Spreadshot [![Build Status](https://travis-ci.org/gkosae/spreadshot.svg?branch=master)](https://travis-ci.org/gkosae/spreadshot) [![Gem Version](https://badge.fury.io/rb/gkosae-spreadshot.svg)](https://badge.fury.io/rb/gkosae-spreadshot) [![Coverage Status](https://coveralls.io/repos/github/gkosae/spreadshot/badge.svg?branch=master)](https://coveralls.io/github/gkosae/spreadshot?branch=master)
A Ruby library for reading spreadsheets with support for multiple backends. This is an attempt to:
  - Provide a uniform interface for reading different formats of spreadsheets.
  - Reduce the amount of code changed when you decide to switch reader backends for any reason

[SmarterCSV](https://github.com/tilo/smarter_csv) (for csv files) and [RubyXL](https://github.com/weshatheleopard/rubyXL) (for xlsx files) backends are provided out of the box.

The name was inspired by the names of [battle chips from Megaman Battle Network 3](https://megaman.fandom.com/wiki/List_of_Mega_Man_Battle_Network_3_Battle_Chips).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'gkosae-spreadshot'
```

And then execute:

    $ bundle

Or install it yourself as:
```bash
$ gem install 'gkosae-spreadshot'
```

## Usage

Require gem with:

```ruby
require 'spreadshot'
```

Read spreadsheet with one of the provided backends
```ruby
csv_reader = Spreadshot::Reader.new(backend: :smarter_csv)

begin
  csv_reader.read(path_to_sample_csv_file) do |row_index, row_data|
    puts "#{row_index} - #{row_data}"
  end
rescue Spreadshot::ReaderError
  # Handle error
end

# A cell_index_to_field_mapper must be provided for the RubyXL backend to map values read from a column to a unique key in the yielded hash
xlsx_reader = Spreadshot::Reader.new(
  backend: :ruby_xl,
  backend_options: {
    cell_index_to_field_mapper: {0 => :h1, 1 => :h2, 2 => :h3, 3 => :h4}
  }
)

begin
  xlsx_reader.read(path_to_sample_xlsx_file) do |row_index, row_data|
    puts "#{row_index} - #{row_data}"
  end
rescue Spreadshot::ReaderError
  # Handle error
end

# SAMPLE OUTPUT
# 2 - {:h1=>11, :h2=>22, :h3=>33, :h4=>44}
# 3 - {:h1=>111, :h2=>222, :h3=>333, :h4=>444}
# 4 - {:h1=>1111, :h2=>2222, :h3=>3333, :h4=>4444}
```

`row_index` begins from 2 because index 1 is used for headers. But if the spreadsheet does not contain headers then the backend can be configured with `headers: false` to begin reading from index 1 (This is only available for the RubyXL backend. To the best of my knowledge SmarterCSV does not support reading files without headers).

```ruby
xlsx_reader = Spreadshot::Reader.new(
  backend: :ruby_xl,
  backend_options: {
    worksheet_index: 1, # Index of the worksheet to read from (defaults to 0, i.e. the first worksheet)
    headers: false,
    cell_index_to_field_mapper: {0 => :h1, 1 => :h2, 2 => :h3, 3 => :h4}
  }
)
begin
  xlsx_reader.read(path_to_sample_xlsx_file) do |row_index, row_data|
    puts "#{row_index} - #{row_data}"
  end
rescue Spreadshot::ReaderError
  # Handle error
end

# SAMPLE OUTPUT
# 1 - {:h1=>11, :h2=>22, :h3=>33, :h4=>44}
# 2 - {:h1=>111, :h2=>222, :h3=>333, :h4=>444}
# 3 - {:h1=>1111, :h2=>2222, :h3=>3333, :h4=>4444}
```

OR<br/><br/>
Read spreadsheet with a custom backend. The custom backend MUST inherit from `Spreadshot::Backends::ReaderBackend` and override `#read`. `#read` MUST yield the index and content(as a Hash) of each row read.<br/><br/>
Example:
```ruby
class SmarterCSVBackend < Spreadshot::Backends::ReaderBackend
  def initialize(options = {})
    options ||= {}
    options[:headers] = true
    super(options)
  end
      
  def read(path_to_spreadsheet)
    SmarterCSV.process(path_to_spreadsheet) do |row|
      current_row_data = row.first
      yield(@current_row_index, current_row_data)
      @current_row_index += 1
    end
  rescue => e
    raise Spreadshot::ReaderError, e.message
  end
end

csv_backend = Spreadshot::Backends::SmarterCSVBackend.new
csv_reader = Spreadshot::Reader.new(backend: csv_backend)
```

Reader backends can also be switched dynamically with the `#set_backend` method
```ruby
reader = Spreadshot::Reader.new(backend: :smarter_csv)

xlsx = Spreadshot::Backends::RubyXLBackend.new(
  worksheet_index: 0, 
  cell_index_to_field_mapper: {0 => :h1, 1 => :h2, 2 => :h3, 3 => :h4}
)
reader.set_backend(xlsx)

#OR

reader.set_backend(
  :ruby_xl,
  {
    headers: false,
    cell_index_to_field_mapper: {0 => :h1, 1 => :h2, 2 => :h3, 3 => :h4}
  }
)
```


## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/gkosae/spreadshot.git

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

CHANGELOG.md
# Changelog
## v0.1.3, 2020-28-07 (Maintenance Release)
- Merged dependabot's security updates

## v0.1.2, 2020-22-03 (Maintenance Release)
- Merged dependabot's security updates
- Integrated Coveralls and updated README with badges
- Integrated TravisCI and updated README with badges

## v0.1.1, 2019-11-01 (Maintenance Release)
- Updated gemspec with CHANGELOG URI and documentation URI
- Corrected RubyXL backend usage instructions in README
- Updated README installation instructions with new gem name
- Changed gem name from 'spreadshot' to 'gkosae-spreadshot' to circumvent rubygems typo protections

## v0.1.0, 2019-10-31 (Feature Release)
- Initial release

spec/errors_spec.rb
RSpec.describe Spreadshot::SpreadshotError do
  it "keeps previous error's backtrace if raised after another error" do
    error = begin
      original_exception = Exception.new
      original_exception.set_backtrace(%W[ foo bar baz ])
      raise original_exception
    rescue Exception
      raise Spreadshot::SpreadshotError.new("test") rescue $!
    end

    expect(error.backtrace).to eq(%W[ foo bar baz ])
  end
end

spec/reader_spec.rb
require 'support/fake_backends'

RSpec.describe Spreadshot::Reader do
  context 'when initialized with a provided backend' do
    let(:supported_backend) { {id: :smarter_csv, options: {}} }
    let(:unsupported_backend) { {id: :solarize, options: {}} }

    it 'raises a Spreadshot::BackendNotFoundError if the provided backend is not recognized' do
      expect { 
        Spreadshot::Reader.new(
          backend: unsupported_backend[:id],
          backend_options: unsupported_backend[:options],
        ) 
      }.to raise_error(Spreadshot::BackendNotFound)
    end
    
    it 'creates an instance of Spreadshot::Reader if the provided backend is recognized' do
      reader = Spreadshot::Reader.new(
        backend: supported_backend[:id],
        backend_options: supported_backend[:options],
      ) 

      expect(reader).to be_a(Spreadshot::Reader)
    end
  end
  

  context 'when initialized with a custom backend' do
    let(:valid_backend) { ValidBackend.new }
    let(:invalid_backend) { InvalidBackend.new }

    it 'raises a Spreadshot::BackendNotFoundError if the custom backend is not a kind of Spreadshot:Backends::ReaderBackend' do
      expect { Spreadshot::Reader.new(backend: invalid_backend) }
        .to raise_error(Spreadshot::BackendNotFound)
    end
    
    it 'creates an instance of Spreadshot::Reader if the provided backend is a kind of Spreadshot:Backends::ReaderBackend' do
      expect(Spreadshot::Reader.new(backend: valid_backend))
        .to be_a(Spreadshot::Reader)
    end
  end

  
  context 'when initialized with a valid backend' do
    let(:valid_backend) { ValidBackend.new }

    it 'expects #read to yield control' do
      reader = Spreadshot::Reader.new(backend: valid_backend)

      expect { |b| reader.read('path_to_sample_spreadsheet', &b) }
        .to yield_control
    end
  end

  describe '#set_backend' do
    context 'when given an invalid backend' do
      let(:provided_backend) { {id: :solarize, options: {}} }
      let(:custom_backend) { InvalidBackend.new }

      it 'raises a Spreadshot::BackendNotFound error' do
        expect { 
          Spreadshot::Reader.new(
            backend: provided_backend[:id],
            backend_options: provided_backend[:options],
          ) 
        }.to raise_error(Spreadshot::BackendNotFound)

        expect { Spreadshot::Reader.new(backend: custom_backend) }
          .to raise_error(Spreadshot::BackendNotFound)
      end
    end
  end
end

spec/spec_helper.rb
require 'coveralls'
Coveralls.wear!

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


require 'spreadshot'

lib/spreadshot/backends/reader_backend.rb
module Spreadshot
  module Backends

    # Base class for all spreadshot reader backends.
    class ReaderBackend

      # @param [Hash] options
      # @option options [Boolean] :headers 
      #   Specifies whether the spreadsheet to be read contains headers
      def initialize(options = {})
        options ||= {}
        @headers = (options.has_key?(:headers)) ? options[:headers] : true
        @current_row_index = @headers ? 2 : 1
      end




      # Reads data from the specified spreadsheet
      # @note 
      #   Must be overriden by subclasses.
      #   Override must raise {Spreadshot::ReaderError Spreadshot::ReaderError} if something goes wrong while reading
      #
      #
      # @param [String] path_to_spreadsheet
      #
      # @yield [row_index, row_data]
      # @yieldparam [Integer] row_index 
      #   The index of the current row being read. The first row has an index of 1
      # @yieldparam [Hash] row_data
      #   A hash representation of the data read from the current row
      def read(path_to_spreadsheet)
        raise NotImplementedError
      end
    end
  end
end

lib/spreadshot/backends/ruby_xl_backend.rb
require 'rubyXL'

module Spreadshot
  module Backends

    # Adapter for the RubyXL Gem ({https://github.com/weshatheleopard/rubyXL})
    class RubyXLBackend < ReaderBackend

      # Returns a new instance of RubyXLBackend (See {ReaderBackend#initialize Spreadshot::ReaderBackends::ReaderBackend})
      #
      # @param [Hash] options
      # @option options [Boolean] :headers
      #   Specifies whether the spreadsheet to be read contains headers. 
      #   Defaults to true
      # @option options [Integer] :worksheet_index
      #   The index of the worksheet to be read. Used for the RubyXL backend.
      #   Defaults to 0
      # @option options [Hash] :cell_index_to_field_mapper
      #   A hash which maps each cell in a row of the spreadsheet to a key in the
      #   hash created for that row. Must be provided
      #
      # @raise [ArgumentError] if cell_index_to_field_mapper option is not provided
      # @raise [ArgumentError] if cell_index_to_field_mapper option is not a Hash
      def initialize(options = {})
        options ||= {}
        super(options)

        begin
          @worksheet_index = Integer(options[:worksheet_index])
        rescue TypeError
          @worksheet_index = 0
        end
        
        begin
          @cell_index_to_field_mapper = Hash(options[:cell_index_to_field_mapper])
        rescue TypeError
          @cell_index_to_field_mapper = options[:cell_index_to_field_mapper]
        end

        raise ArgumentError, 'cell_index_to_field_mapper is invalid' unless @cell_index_to_field_mapper.is_a?(Hash)
        raise ArgumentError, 'cell_index_to_field_mapper cannot be empty' if @cell_index_to_field_mapper.empty?
      end
      



      # Reads data from the specified spreadsheet
      # (See {ReaderBackend#read Spreadshot::ReaderBackends::ReaderBackend})
      #
      # @param [String] path_to_spreadsheet
      #
      # @yield [row_index, row_data]
      # @yieldparam [Integer] row_index 
      #   The index of the current row being read. The first row has an index of 1
      # @yieldparam [Hash] row_data
      #   A hash representation of the data read from the current row
      #
      # @raise [Spreadshot::ReaderError]
      def read(path_to_spreadsheet)
        worksheet = RubyXL::Parser.parse(path_to_spreadsheet)[@worksheet_index]
        headers = @headers
  
        worksheet.each do |row|
          if headers
            headers = false
            next
          end
    
          cell_index = 0
          current_row_data = {}  
    
          row && row.cells.each do |cell|
            val = cell && cell.value
    
            if @cell_index_to_field_mapper[cell_index]
              current_row_data[@cell_index_to_field_mapper[cell_index]] = val
            end
    
            cell_index += 1
          end
          
          yield(@current_row_index, current_row_data)
          @current_row_index += 1
        end
  
        return nil
      rescue => e
        raise Spreadshot::ReaderError, e.message
      end
    end
  end
end

lib/spreadshot/backends/smarter_csv_backend.rb
require 'smarter_csv'

module Spreadshot
  module Backends

    # Adapter for the SmarterCSV Gem ({https://github.com/tilo/smarter_csv})
    class SmarterCSVBackend < ReaderBackend

      # Returns a new instance of SmarterCSVBackend (See {ReaderBackend#initialize Spreadshot::ReaderBackends::ReaderBackend})
      # @param [Hash] options
      # @note Ignores :header option
      def initialize(options = {})
        options ||= {}
        options[:headers] = true
        super(options)
      end
      



      # Reads data from the specified CSV file
      # (See {ReaderBackend#read Spreadshot::ReaderBackends::ReaderBackend})
      #
      # @param [String] path_to_spreadsheet
      #
      # @yield [row_index, row_data]
      # @yieldparam [Integer] row_index 
      #   The index of the current row being read. The first row has an index of 1
      # @yieldparam [Hash] row_data
      #   A hash representation of the data read from the current row
      #
      # @raise [Spreadshot::ReaderError]
      def read(path_to_spreadsheet)
        SmarterCSV.process(path_to_spreadsheet) do |row|
          current_row_data = row.first
          yield(@current_row_index, current_row_data)
          @current_row_index += 1
        end
      rescue => e
        raise Spreadshot::ReaderError, e.message
      end
    end
  end
end

lib/spreadshot/reader.rb
module Spreadshot
  class Reader
    
    # @param [Hash] options
    # @option options [Symbol, Spreadshot::ReaderBackend] :backend 
    #   The reader backend to use or the id if one of the provided backends
    # @option options [Hash] :backend_options
    #   Required if one of the provided backends is used
    #   * :headers (Boolean) [true] Specifies whether the spreadsheet to be read contains headers. Will be ignored if the backend doesn't provide this option (E.g. SmarterCSV expects headers)
    #   * :worksheet_index (Integer) [0] The index of the worksheet to be read. Used for the RubyXL backend
    #   * :cell_index_to_field_mapper (Hash<Integer, [String, Symbol]>) A hash which maps each cell in a row of the spreadsheet to a key in the result hash created for that row. Must be provided for the RubyXL backend
    #
    # @example
    #   reader = Spreadshot::Reader.new(backend: :smarter_csv)
    #
    #   reader = Spreadshot::Reader.new(backend: :ruby_xl, worksheet_index: 0, cell_index_to_field_mapper: {0 => :H1, 1 => :H2, 2 => :H3, 3 => :H4})
    #
    #   xlsx_backend = Spreadshot::Backends::RubyXLBackend.new(worksheet_index: 0, cell_index_to_field_mapper: {0 => :H1, 1 => :H2, 2 => :H3, 3 => :H4})
    #   reader = Spreadshot::Reader.new(backend: xlsx_backend)
    def initialize(options)
      set_backend(options[:backend], options[:backend_options])
    end




    # Set the backend to use
    #
    # @param [Symbol, Spreadshot::Backends::ReaderBackend] backend 
    #   The new backend or id of one of the provided backends
    # @param [Hash] backend_options
    #   Required if one of the provided backends is used
    # @option backend_options [Boolean] :headers
    #   Specifies whether the spreadsheet to be read contains headers. 
    #   Will be ignored if the backend doesn't provide this option (E.g. SmarterCSV expects headers)
    #   Defaults to true
    # @option backend_options [Integer] :worksheet_index
    #   The index of the worksheet to be read. Used for the RubyXL backend.
    #   Defaults to 0
    # @option backend_options [Hash<Integer, [Symbol, String]>] :cell_index_to_field_mapper
    #   A hash which maps each cell in a row of the spreadsheet to a key in the 
    #   result hash created for that row. Used for the RubyXL backend.
    #   Must be provided
    #
    # @raise [Spreadshot::BackendNotFound]
    #   If no backend is provided or if one of the provided backends is not specified
    #
    # @example
    #   reader = Spreadshot::Reader.new(backend: :ruby_xl, worksheet_index: 0, cell_index_to_field_mapper: {0 => :H1, 1 => :H2, 2 => :H3, 3 => :H4})
    #   csv_backend = Spreadshot::Backends::SmarterCSVBackend.new
    #   reader.set_backend(csv_backend)
    def set_backend(backend, backend_options = {})
      @backend = (backend.is_a?(Spreadshot::Backends::ReaderBackend)) ? backend : build_backend(
        backend,
        backend_options
      )

      raise Spreadshot::BackendNotFound if @backend.nil?
    end




    # Reads data from the specified spreadsheet
    #
    # @param [String] path_to_spreadsheet
    #
    # @yield [row_index, row_data]
    # @yieldparam [Integer] row_index
    #   The index of the current row being read. The first row has an index of 1
    # @yieldparam [Hash] row_data
    #   A hash representation of the data read from the current row
    #
    # @note This method delegates actual reading to the backend
    #
    # @example
    #   reader = Spreadshot::Reader.new(backend: :ruby_xl, worksheet_index: 0, cell_index_to_field_mapper: {0 => :H1, 1 => :H2, 2 => :H3, 3 => :H4})
    #   reader.read('spreadshot_test.xlsx'){|row_index, row_data| puts "#{row_index} - #{row_data}"}
    #   
    #   Sample output
    #   2 - {:H1=>11, :H2=>22, :H3=>33, :H4=>44}
    #   3 - {:H1=>111, :H2=>222, :H3=>333, :H4=>444}
    #   4 - {:H1=>1111, :H2=>2222, :H3=>3333, :H4=>4444}

    def read(path_to_spreadsheet)
      @backend.read(path_to_spreadsheet) {|row_index, row_data| yield(row_index, row_data)}
    end




    private

    # Creates a reader backend using the backend (id) and options/backend_options provided in #initialize or #set_backend
    #
    # @param [Symbol] backend_id the id(key) of the desired backend
    # @param [Hash] backend_options 
    #   options to build the backend with
    # @option backend_options [Boolean] :headers
    #   Specifies whether the spreadsheet to be read contains headers. 
    #   Will be ignored if the backend doesn't provide this option (E.g. SmarterCSV expects headers)
    #   Defaults to true
    # @option backend_options [Integer] :worksheet_index
    #   The index of the worksheet to be read. Used for the RubyXL backend.
    #   Defaults to true
    # @option backend_options [Hash<Integer, [Symbol, String]>] :cell_index_to_field_mapper
    #   A hash which maps each cell in a row of the spreadsheet to a key in the 
    #   result hash created for that row. Used for the RubyXL backend.
    #   Must be provided
    #
    # @return [Spreadshot::Backends::ReaderBackend]
    def build_backend(backend_id, backend_options = {})
      backend_options ||= {}
      return provided_backends[backend_id].call(backend_options)
    rescue NoMethodError
      return nil
    end




    # The provided backends
    # 
    # @return [Hash]
    def provided_backends
      {
        smarter_csv: ->(backend_options){ Spreadshot::Backends::SmarterCSVBackend.new(backend_options) },
        ruby_xl: ->(backend_options){ Spreadshot::Backends::RubyXLBackend.new(backend_options) },
      }
    end
  end
end